clash-lib/src/Clash/Normalize/Transformations/Inline.hs

```diff
@@ -379,6 +382,45 @@ inlineCleanup (TransformContext is0 _) (Letrec binds body) = do
 inlineCleanup _ e = return e
 {-# SCC inlineCleanup #-}
 
+-- | Takes a binding and collapses its term if it is a noop
+inlineIdentity :: HasCallStack => NormRewrite
+inlineIdentity _ (Letrec binds body) = do
+  (untouched,collapsed) <- partitionEithers <$> mapM identityAnalysis binds
+  if null collapsed
+    then return (Letrec binds body)
+    else changed (Letrec (untouched ++ collapsed) body)
+inlineIdentity _ e = return e
+
+identityAnalysis :: HasCallStack => (Id,Term) -> NormalizeSession (Either (Id,Term) (Id,Term))
+identityAnalysis old@(iD,collectArgs -> (term,partitionEithers -> (termArgs,typeArgs))) =
+  runIdentityAnalysis $ do
+    Prim primInfo <- return term
```

Comment:
As this is limited to things where `collectArgs` has a `Prim primInfo` in the function position, we could match for this earlier. The line is already quite long so maybe [pattern guards](https://wiki.haskell.org/Pattern_guard) would be more readable, e.g.

```haskell
identityAnalysis old@(collectArgs -> (term, args))
  | Prim primInfo <- term
  , (termArgs, typeArgs) <- partitionEithers args
  = do ...
```

Reply:
Point taken, we can abort here earlier. 
I'm not a big fan of those pattern-matching + pattern-guards + fallback clauses, so I am going to propose an alternative.